src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs

```diff
@@ -653,18 +686,35 @@ private bool StartTag(Stack<Tuple<HtmlSymbol, SourceLocation>> tags, IDisposable
         {
             // If we're at text, it's the name, otherwise the name is ""
             HtmlSymbol tagName;
-            if (At(HtmlSymbolType.Text))
+            HtmlSymbol bangSymbol = null;
+            HtmlSymbol potentialTagNameSymbol;
+
+            if (At(HtmlSymbolType.Bang))
```

Comment:
Investigate rewriting Symbols to contain `!p` for the `tagName`. This would enable correct error cases for:

```
@{
<!p>
</p>
}
```